patches/server/1055-Add-required-plugins-config-option.patch

```diff
@@ -0,0 +1,82 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Tamion <70228790+notTamion@users.noreply.github.com>
+Date: Sat, 25 Nov 2023 23:15:22 +0100
+Subject: [PATCH] Add required-plugins config option
+
+
+diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
+index 08b1aab5d37a56dc42542ce15ba1f7ccd1b08400..71543d2bf811aabfcfda57b7ac9dd91c70788ff5 100644
+--- a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
++++ b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
+@@ -211,6 +211,13 @@ class PaperPluginInstanceManager {
+         if (!(plugin instanceof JavaPlugin javaPlugin)) {
+             throw new IllegalArgumentException("Only expects java plugins.");
+         }
++        // Paper start
++        if (!this.server.isStopping() && !this.server.isReloading() && io.papermc.paper.configuration.GlobalConfiguration.get().misc.requiredPlugins.contains(plugin.getName())) {
++            this.server.getLogger().severe("The required plugin " + plugin.getName() + " got disabled. Stopping Server.");
++            this.server.shutdown();
++            return;
```

Comment:
Is any weirdness caused by starting the shutdown here when this disable call could be in disablePlugins which is a loop over all the plugins, calling disable on each of them? I haven't tested it, it just might be a concern.

Reply:
honestly i think just removing the shutdown here is best. maybe logging that there was an attempt to disable the plugin or making the method return a boolean depending on if the plugin was disabled succesfully